File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    private final String[] KEY = {
            "",
            "",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
        };
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        if(digits == null || digits.length()==0) return res;
        StringBuilder sb = new StringBuilder();
        backtrack(digits, 0, sb, res);
        return res;
    }

    public void backtrack(String digits, int i, StringBuilder sb, List<String> res){
        if(i==digits.length()){
            res.add(sb.toString());
            return;
        }

        String letters = KEY[digits.charAt(i)-'0'];
        for(char c : letters.toCharArray()){
            sb.append(c);
            backtrack(digits, i+1, sb, res);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        digitToChar = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }
        res = []
        def backtrack(i, cur):
            if i >= len(digits):
                res.append(cur)
                return
            
            for c in digitToChar[digits[i]]:
                backtrack(i+1, cur + c)
        backtrack(0, "")
        return res